drv/riscv-plic/build.rs

```diff
@@ -0,0 +1,163 @@
+    writeln!(file, "type Priority = plic::Priority<PLIC_PRIORITY_BITS>;")
+        .unwrap();
+
+    /*
+    let mut task_set: HashSet<String> = HashSet::new();
+    for task in TASK_CONFIG.tasks.iter() {
+        if task_set.insert(task.to_string()) == true {
+            writeln!(file, "task_slot!({}, {});", task.to_ascii_uppercase(), task);
+    }
+    */
+
+    use abi::{InterruptNum, InterruptOwner, TaskId};
+    let fmt_irq_task = |v: Option<&(InterruptNum, (TaskId, u32))>| {
+        match v {
+            Some((irq, owner)) => format!(
+                "(userlib::InterruptNum({}), (TaskId({}), 0b{:b})),",
+                irq.0, owner.0.0, owner.1
+            ),
+            None => "(userlib::InterruptNum::invalid(), userlib::InterruptOwner::invalid()),"
+                .to_string(),
+        }
+    };
+
+    let fmt_task_irq = |v: Option<&(InterruptOwner, Vec<InterruptNum>)>| {
+        match v {
+            Some((owner, irqs)) => format!(
+                "(userlib::InterruptOwner {{ task: {}, notification: 0b{:b} }}, &[{}]),",
+                owner.task, owner.notification,
+                irqs.iter()
+                    .map(|i| format!("userlib::InterruptNum({})", i.0))
+                    .collect::<Vec<_>>()
+                    .join(", ")
+            ),
+            None => {
+                "(TaskId::IndexMask - 1, &[]),"
+                    .to_string()
+            }
+        }
+    };
+
+    println!("{}", &env::var("HUBRIS_TASK_ID_MAP")?);
+    let task_id_map: std::collections::BTreeMap<String, u16> =
+        ron::de::from_str(&env::var("HUBRIS_TASK_ID_MAP")?)?;
+
+    let mut irq_task_map: Vec<(InterruptNum, (TaskId, u32))> = Vec::new();
+    for (i, irq) in TASK_CONFIG.ints.into_iter().enumerate() {
+        let task: String = TASK_CONFIG.tasks[i].to_string();
+        let task_id: TaskId = match task_id_map.get(&task) {
+            Some(id_num) => TaskId(*id_num),
+            None => panic!("Error: no matching task ID for task {}", task),
+        };
+        let notification: u32 = TASK_CONFIG.notification[i];
+        irq_task_map.push((InterruptNum(*irq as u32), (task_id, notification)));
+    }
+
+    let mut per_task_irqs: HashMap<InterruptOwner, Vec<InterruptNum>> =
+        HashMap::new();
+    for (i, task) in TASK_CONFIG.tasks.iter().enumerate() {
+        let notification: u32 = TASK_CONFIG.notification[i];
+        let irq: InterruptNum = InterruptNum(TASK_CONFIG.ints[i] as u32);
+        let task: String = task.to_string();
+        let task_id: TaskId = match task_id_map.get(&task) {
+            Some(id_num) => TaskId(*id_num),
+            None => panic!("Error: no matching task ID for task {}", task),
+        };
+        let owner: InterruptOwner = InterruptOwner {
+            task: task_id.index() as u32,
+            notification: notification,
+        };
+        per_task_irqs.entry(owner).or_default().push(irq);
```

Comment:
Since `per_task_irqs` is building a vector of interrupt numbers, does it mean that a single task can request same notification for different interrupt numbers?

Reply:
Yeah, there's nothing stopping a task from doing this.

Reply:
But in that case how would a task tell the plic driver which interrupt it really wants to mask/unmask/complete?

Reply:
Presumably it'll do the operation on all of them; that's how it's done in the kernel (see the [irq_control](https://github.com/rivosinc/hubris/blob/502e5ae9797f846d65830742b0d6a99719513db7/sys/kern/src/syscalls.rs#L689-L719) syscall)